### PR TITLE
Added support for pre_format annotation

### PR DESCRIPTION
**Pull request by chiragsanghvi**
This PR adds support for pre_format annotation as mentioned in the issue https://github.com/informatics-isi-edu/ermrestjs/issues/299. 

The format for the printf function is as follows
```js

/**
      * printf function allows you to format strings using POSIX statndard
      *
      * %[flags][width][.precision]type
      *
      * flags field: The Flags field can be zero or more (in any order) of:
    
           - (minus) Left-align the output of this placeholder. (The default is to right-align the output.). 
                     This is used in conjunction with width field.
           + (plus)  Prepends a plus for positive signed-numeric types. positive = +, negative = -
                     (The default doesn't prepend anything in front of positive numbers.)
           0 (zero)  When the 'width' option is specified, prepends zeros for numeric types. (The default prepends spaces.)
                     For example, printf("%2X",3) produces  3, while printf("%02X",3) produces in 03.
           ' (quote) It separates numeric types by thousands using a comman(,) or other characters accroding to localization 
           #{char}   An optional padding specifier that says what character to use for padding (if specified). 
                     Possible values are any other character precedeed by a # (hash). print("%#_5d",10) produces ___10
    
      * Width field: The Width field specifies a minimum number of characters to output, and is typically used to pad fixed-width fields, 
                     where the fields would otherwise be smaller, although it does not cause truncation of oversized fields.
                     For example, printf("%*d", 5, 10) will result in "   10" being printed, with a total width of 5 characters.
                     Though not part of the width field, a leading zero is interpreted as the zero-padding flag mentioned above, 
                     and a negative value is treated as the positive value in conjunction with the left-alignment - flag also mentioned above.
                     When used with the j (JSON) type specifier, the padding length specifies the tab size used for indentation.
      
      * Precision field: The Precision field usually specifies a maximum limit on the output, depending on the particular formatting type.
                    For floating point numeric types, it specifies the number of digits to the right of the decimal point that the output should be rounded.
                    For the string type, it limits the number of characters that should be output, after which the string is truncated.
                    For example, printf("%.3s", 3.42134) will result in 3.421 being printed.
        
      * Type field: It can be any of:
                    % — yields a literal % character
                    b — yields an integer as a binary number
                    c — yields an integer as the character with that ASCII value
                    d or i — yields an integer as a signed decimal number
                    e — yields a float using scientific notation
                    u — yields an integer as an unsigned decimal number
                    f — yields a float as is; see notes on precision above
                    o — yields an integer as an octal number
                    s — yields a string as is
                    t — yields true or false
                    X — yields an integer as a hexadecimal number (upper-case)
                    j — yields a JavaScript object or array as a JSON encoded string
      
      * Named arguments:
                    Format strings may contain replacement fields rather than positional placeholders. Instead of referring to a certain argument, 
                    you can now refer to a certain key within an object. Replacement fields are surrounded by rounded parentheses - `(` and `)` - 
                    and begin with a keyword that refers to a key:

                        var user = { name: 'Dolly' }
                        sprintf('Hello %(name)s', user) // Hello Dolly

                    Keywords in replacement fields can be optionally followed by any number of keywords or indexes:

                        var users = [
                            {name: 'Dolly'},
                            {name: 'Molly'},
                            {name: 'Polly'},
                        ]
                        sprintf('Hello %(users[0].name)s, %(users[1].name)s and %(users[2].name)s', {users: users}) // Hello Dolly, Molly and Polly

                    Note: mixing positional and named placeholders is not (yet) supported
                    

       * @param{Object} options A javascript object that contains format property and optional bool_true_value
                                and bool_false_value. These will be used to replace true and false in `t` type fields.
       * @param{String|Number|Object} value This would be the value that needs to be formatted.

       * @returns {String}
       **/
```

### CODE REVIEW

File: js/core.js [hunk 1]
```diff
@@ -1719,6 +1722,16 @@
                 return this._getNullValue(context);
             }
 
+            if (this.annotations.contains(module._annotations.COLUMN_DISPLAY)) {
```
Comment (RFSH) [inline js/core.js]: This is not necessary. Please remove it.

File: js/core.js [hunk 2]
```diff
@@ -2039,6 +2047,8 @@
                 }
 
                 this._display[context] = {
+                    "isPreformat": (typeof annotation.pre_format === 'string'),
```
Comment (RFSH) [inline js/core.js]: This seems wrong. You should get the format before this line and in here check if that is string or not.
```javascript
var preformatConfig = typeof annotation.pre_format === 'object') ? ((typeof annotation.pre_format.format === 'string') ? annotation.pre_format : null ) : null;

this._display[context] = {
  "isPreformat": (typeof preformatConfig === 'string'),
  "preformatConfig": preformatConfig,
   //..
}
```

**Comment by RFSH** [comment]
Please add some test cases that are testing the formatValue function and the fact that is actually using the annotations (Current implementation is not using the annotation because of the bug that I mentioned in the comments). You're only testing the internal function (`printf`). You can add them to the reference_values spec that you have created before, or any other specs that you think is more appropriate.

**Comment by chiragsanghvi** [comment]
@RFSH I have added testcases for the annotation to be used in reference.

File: js/core.js [hunk 3]
```diff
@@ -2046,9 +2044,16 @@
                     columnOrder = annotation.column_order;
                 }
 
+                if (typeof annotation.pre_format === 'object') {
+                    if (typeof annotation.pre_format.format !== 'string') {
```
Comment (RFSH) [inline js/core.js]: You don't need to throw an error, just ignore the annotation as if it didn't exist. You can just do a `console.log`.

Comment (chiragsanghvi) [reply js/core.js]: Alright I have removed it.

File: js/core.js [hunk 4]
```diff
@@ -1723,8 +1726,16 @@
...
+            if (data === undefined || (data === null && this.type.name.indexOf('json') === -1)) {
                 return this._getNullValue(context);
+            } else if (data === null && this.type.name.indexOf('json') !== 0) {
+                return data;
+            }
+
+            var display = this.getDisplay(context);
+
+            if (display.isPreformat) {
+                return module._printf(display.preformatConfig, data);
```
Comment (RFSH) [inline js/core.js]: This function might throw an error which we don't want to raise to the users. Wrap this in a try-catch to ignore the pre_format annotation in that case.

**Pull request merged**